### PR TITLE
NEW Allow files to have a .tmpl extension which is stripped

### PR DESCRIPTION
**Pull request by dhensby**
Allows a fix for https://github.com/silverstripe/recipe-cms/issues/9

--

Added:
 - Tests, travis integration, test namespaces and autoloading
 - Added APIs to better facilitate mocking (such as `file_exists`, `file_get_contents`, `copy`, etc)
 - Support for recipe files to end in `.tmpl` so that they don't get picked up by IDEs as first class classes

### CODE REVIEW

**Comment by tractorcow** [comment]
@dhensby good idea but it won't work if you `composer create-project silverstripe/recipe-cms`. Can you make sure it renames base files as well?

**Comment by dhensby** [comment]
@tractorcow this appeared to be the canonical source of mapping from the composer meta data to the filename and I couldn't (and can't now) find anywhere referencing `postCreate` hooks / logic.

Any pointers where to look for that?

File: src/RecipeInstaller.php [hunk 1]
```diff
@@ -44,6 +44,10 @@ protected function installProjectFiles($recipe, $sourceRoot, $destinationRoot, $
         $any = false;
         foreach($fileIterator as $path => $info) {
             $destination = $destinationRoot . substr($path, strlen($sourceRoot));
+            $extension = pathinfo($destination, PATHINFO_EXTENSION);
+            if ($extension === 'tmpl') {
+                $destination = substr($destination, -5);
```
Comment (tractorcow) [inline src/RecipeInstaller.php]: Did you mean `substr($destination, 0, -5)`? `$destination` would end up just being `.tmpl`.

**Comment by tractorcow** [comment]
Check out this bit of code:


```php
public static function getSubscribedEvents()
    {
        return [
            'post-create-project-cmd' => 'cleanupProject',
            'post-package-update' => 'installPackage',
            'post-package-install' => 'installPackage',
        ];
    }
```

You would want to add your hook in the `cleanupProject` hook. Unfortunately you may need to do a bit of refactoring in RecipeInstaller; I suggest a new method designed for renaming methods in the base recipe.

The logic for this method would look VERY similar to `RecipeInstaller::installProjectFiles()`, except instead of installing files it'll just rename root project files.

Let me know if you need any more help. ;P

**Comment by tractorcow** [comment]
This check

```
// Add file to installed (even if already exists)
            if (!in_array($relativePath, $installedFiles)) {
```

Should we loosen this check so that it matches path with / without the tml suffix? Just to ensure backwards compatibility.

File: src/RecipeInstaller.php (hunk 1 above)
Comment (dhensby) [reply src/RecipeInstaller.php]: whoops - fixed

**Comment by dhensby** [comment]
@tractorcow `cleanupProject` doesn't appear to inspect the values of files that are defined in `extra` meta data. it just unsets the values... I'm not sure how the `.tmpl` logic would fit in here? sorry if I'm being dense.

https://github.com/silverstripe/recipe-plugin/blob/master/src/RecipePlugin.php#L89-L110

**Comment by dhensby** [comment]
> Should we loosen this check so that it matches path with / without the tml suffix? Just to ensure backwards compatibility

This is currently the case. We normalise the `.tmpl` path early so that comparisons are done as if it was always named without it. Thus it's written to the composer.json without the `.tmpl` (if it had one) and compared without it too, I've added a test case to cover this under `testInstallProjectFilesWithoutTmplExtension`

**Comment by tractorcow** [comment]
> @tractorcow cleanupProject doesn't appear to inspect the values of files that are defined in extra meta data. it just unsets the values... I'm not sure how the .tmpl logic would fit in here? sorry if I'm being dense.


You do it before clearing them. The logic is:

 - Install project
   - Install each library in project
     - plugin installs files for each library
 - read composer.json in cleanupProject:
   - For each extra, rename `.tmpl`
   - then unset it
   
> We normalise the .tmpl path early

Good call just making sure. :)

File: src/RecipeInstaller.php [hunk 2]
```diff
@@ -32,23 +41,23 @@ public function __construct(IOInterface $io, Composer $composer) {
...
         // Load all project files
         $fileIterator = $this->getFileIterator($sourceRoot, $filePatterns);
         $any = false;
         foreach($fileIterator as $path => $info) {
             $destination = $destinationRoot . substr($path, strlen($sourceRoot));
-            $extension = pathinfo($destination, PATHINFO_EXTENSION);
-            if ($extension === 'tmpl') {
-                $destination = substr($destination, -5);
+            $destinationExt = pathinfo($destination, PATHINFO_EXTENSION);
+            if ($destinationExt === 'tmpl') {
```
Comment (tractorcow) [inline src/RecipeInstaller.php]: Do you need this condition twice? They both compare the same string that comes from $path? It's hard to follow.
$path -> $destination -> check ext
$path -> $relativePath -> check ext

**Comment by maxime-rainville** [comment]
Is this still blocked? Looks like all of @tractorcow feedback has been addressed?

**Comment by dhensby** [comment]
> Is this still blocked? Looks like all of @tractorcow feedback has been addressed?

it just doesn't work. I think it's actually really difficult to do :/

**Comment by maxime-rainville** [comment]
:disappointed: Shall we close the PR then? Doesn't look like it's going to be addressed in the near future.

**Comment by dhensby** [comment]
Yeah - I think I got as far as I could and wasn't getting anywhere when I last worked on it. It's a big shame, though, as it's a big pain when IDEs pick up these files :(
